### ISSUE TITLE
As long as the payment isn't confirmed the purchase shouldn't be part of the result of restorePurchases

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#124994]
_The following was reported to our security alias. It does not seem to me to be a security vulnerability so I am filing it as a normal issue. -@Hixie_

----

## Steps to Reproduce

I use in_app_purchase: 3.1.5 and test on an emulator inside Android Studio.

1. Buy a product with a slow test card. E.g.: "Slow test card, declines after a few minutes"
2. Call InAppPurchase.instance.restorePurchases() within the next minute

**Expected results:**
As long as the payment isn't confirmed the purchase shouldn't be part of the result of restorePurchases. Or at least should have the state: PurchaseStatus.pending

**Actual results:**
Purchase is returned when calling restorePurchases and [state] is PurchaseStatus.restored, [pendingCompletePurchase] is true

With these values I'm not able to say if the payment is confirmed or not. Giving access to a product can therefore be exploided by using a delayed, declining credit card and call restorePurchases.

(Stopping the delivery of the product when PurchaseStatus.restored && pendingCompletePurchase == true
isn't an option, because I want to be able to restore a purchase when the user behaves like this: Purchases a product and immediately closes the app)

It is possible to set the correct `PurchaseStatus` by hand. I call the following function inside `InAppPurchase.instance.purchaseStream.listen` for each `PurchaseDetails`:

```dart
static void _bugfixSetPending(PurchaseDetails purchaseDetails) {
  if (Platform.isAndroid &&
      purchaseDetails.status == PurchaseStatus.restored) {
    if (purchaseDetails is GooglePlayPurchaseDetails) {
      final state = purchaseDetails.billingClientPurchase.purchaseState;
      if (state == PurchaseStateWrapper.pending) {
        purchaseDetails.status = PurchaseStatus.pending;
      }
    } else {
      logger.w('Failed to access google purchase details.');
    }
  }
}
```

**Comment by WasaLordy** [comment]
Thanks for following up on the problem.
(I still think it's related to security. I guess there are some app's where you could get products for free)  ;)

**Comment by kstoykov** [comment]
I got exactly the same issue :(

**Comment by WasaLordy** [comment]
Agree... But for now I hope my function [_bugfixSetPending] helps you.
As for as I know it works perfectly :)

**Comment by serhii-k** [comment]
Hello guys,

I'm glad this issue is already opened, and it was even created by @Hixie. Thank you!

I have a critical add-on on this.

This bug is still present:
in_app_purchase 3.1.7
in_app_purchase_android 0.3.0+10

**When the actual cancelation happens after a few minutes, no event is being added to the InAppPurchase.instance.purchaseStream.**

**Important!**

The same issue occurs with "Slow test card, **approves** after a few minutes".

In this case:
1.  At first we're getting the 'pending' event on InAppPurchase.instance.purchaseStream. That's okay.
2. But then during minutes we don't receive any event about the purchase is completed. Even though the payment is already accepted.
3. Until we call ourselves restorePurchases().

I guess it started to happen with transition to the Google Play Billing Library 5 (This issue was already present with in_app_purchase_android around 0.2.5).

**Comment by serhii-k** [comment]
As a temporary workaround, in addition to @WasaLordy 's solution. Whenever we're getting any pending (a real or a "fixed" pending) PurchaseDetails on the InAppPurchase.instance.purchaseStream, we need to start a timer to poll the restorePurchases() again.

**Comment by WasaLordy** [comment]
Hi @serhii-k 
I tested your described behaviour today and I used the same versions.
At least on my emulator I can agree that the "slow and declining test card" doesn't trigger any event after "PurchaseStatus.pending". I waited 16 minutes. Also no email is send from google to inform about the purchase error.

But for the "slow approves test card" I do get a second event after "PurchaseStatus.pending".
After 63 seconds I got "PurchaseStatus.purchased" and 5 seconds later the email from google with "Thank you" reached my mailbox.

**Comment by serhii-k** [comment]
Hi @WasaLordy
Yes, you are absolutely right. Thank you!
I have just tried it once again after reading your message, and it does send the "PurchaseStatus.purchased" event for the "slow approves test card".
